Description: Normalize the paths to distribution searchplugins so that
 associated metadata is not lost on upgrades (when the install location changes)
Author: Alexander Sack <asac@ubuntu.com>
Bug: https://bugzilla.mozilla.org/show_bug.cgi?id=534663
Forwarded: no
---
 toolkit/components/search/nsSearchService.js |    9 +++++++++
 1 file changed, 9 insertions(+)

Index: firefox-trunk-41.0~a1~hg20150706r251448/toolkit/components/search/nsSearchService.js
===================================================================
--- firefox-trunk-41.0~a1~hg20150706r251448.orig/toolkit/components/search/nsSearchService.js	2015-07-06 14:38:25.502892340 +0100
+++ firefox-trunk-41.0~a1~hg20150706r251448/toolkit/components/search/nsSearchService.js	2015-07-06 14:38:25.466892238 +0100
@@ -50,6 +50,7 @@
 const MODE_TRUNCATE = 0x20;
 
 // Directory service keys
+const NS_XPCOM_CURRENT_PROCESS_DIR = "XCurProcD";
 const NS_APP_SEARCH_DIR_LIST  = "SrchPluginsDL";
 const NS_APP_DISTRIBUTION_SEARCH_DIR_LIST = "SrchPluginsDistDL";
 const NS_APP_USER_SEARCH_DIR  = "UsrSrchPlugns";
@@ -3896,6 +3897,13 @@
       if (!file.isFile() || file.fileSize == 0 || file.isHidden())
         continue;
 
+      var appdirstr = getDir (NS_XPCOM_CURRENT_PROCESS_DIR).path
+                      + "/distribution/searchplugins";
+
+      // normalize distribution searchplugins if they are following links
+      if (file.path.indexOf(appdirstr) == 0)
+        file.normalize();
+
       var fileURL = NetUtil.ioService.newFileURI(file).QueryInterface(Ci.nsIURL);
       var fileExtension = fileURL.fileExtension.toLowerCase();
       var isWritable = isInProfile && file.isWritable();
